net/portmapper/probe.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package portmapper
+
+import (
+	"context"
+	"time"
+
+	"tailscale.com/syncs"
+)
+
+// Prober periodically pings the network and checks for port-mapping services.
+type Prober struct {
+	// stop will stop the prober
+	stop func()
+
+	// Each of the SubResults below is intended to expose whether a specific service is available
+	// for use on a client, and the most recent seen time. Should not be modified externally, and
+	// will be periodically updated.
+
+	// PMP stores the result of probing pmp services and is populated by the prober.
+	PMP syncs.WaitableResult
+	// PCP stores the result of probing pcp services and is populated by the prober.
+	PCP syncs.WaitableResult
+	// UPnP stores the result of probing pcp services and is populated by the prober.
+	UPnP syncs.WaitableResult
+}
+
+// newProber will start a prober if it does not exist on the given portmapping client.
+// Should be called when the client's lock is held.
+func (c *Client) newProber(ctx context.Context) {
+	stop := false
+	p := &Prober{
+		PMP:  syncs.NewWaitableResult(),
+		PCP:  syncs.NewWaitableResult(),
+		UPnP: syncs.NewWaitableResult(),
+		stop: func() { stop = true },
```

Comment:
I think it's fine for this to be racey, that did occur to me while writing, but since it's only written to and read from 1 spot, we're fine with it running for 1 more iteration. either way, I've changed it to an AtomicBool so it won't have any racey-ness